src/functional/ffront/decorator.py

```diff
@@ -191,37 +198,57 @@ def _validate_args(self, *args, **kwargs) -> None:
         if kwargs:
             raise NotImplementedError("Keyword arguments are not supported yet.")
 
-    def __call__(self, *args, offset_provider, **kwargs) -> None:
+    def __call__(self, *args, offset_provider, backend=None, **kwargs) -> None:
         self._validate_args(*args, **kwargs)
 
         # extract size of all field arguments
-        size_args = []
+        size_args: list[Optional[tuple[int, ...]]] = []
+        rewritten_args = list(args)
         for param_idx, param in enumerate(self.past_node.params):
+            if isinstance(param.type, ct.ScalarType):
+                rewritten_args[param_idx] = ConstantField(args[param_idx])
             if not isinstance(param.type, ct.FieldType):
                 continue
+            if args[param_idx].array is None:
+                size_args.append(None)
+                continue
             for dim_idx in range(0, len(param.type.dims)):
                 size_args.append(args[param_idx].shape[dim_idx])
 
-        backend = self.backend if self.backend else DEFAULT_BACKEND
+        if not backend:
+            backend = self.backend if self.backend else DEFAULT_BACKEND
 
         execute_fencil(
-            self.itir, *args, *size_args, **kwargs, offset_provider=offset_provider, backend=backend
+            self.itir,
+            *rewritten_args,
+            *size_args,
+            **kwargs,
+            offset_provider=offset_provider,
+            backend=backend,
         )
 
 
 def program(
-    definition: types.FunctionType, externals: Optional[dict] = None, backend: Optional[str] = None
-) -> Program:
-    """
-    Generate an implementation of a program from a Python function object.
+    definition: Optional[types.FunctionType] = None,
+    *,
+    externals: Optional[dict] = None,
+    backend: Optional[str] = None,
+) -> Callable[[types.FunctionType], Program] | Program:
+    def program_inner(definition: types.FunctionType) -> Program:
+        """
```

Comment:
Docstring should be at the level of the outer function.

Reply:
✅